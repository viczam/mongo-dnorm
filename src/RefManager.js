import { operations, types } from './constants';
import RefStore from './RefStore';
import Events from './Events';
import set from 'lodash/set';

export default class RefManager {
  constructor(db) {
    this.db = db;
    this.events = new Events();
    this.references = new RefStore();
  }

  add(rawConfig) {
    const config = RefStore.parseRefConfig(rawConfig);

    this.references.add(config);

    if (config.syncOn.includes(operations.update)) {
      this.onUpdate(config);
    }

    if (config.syncOn.includes(operations.remove)) {
      this.onRemove(config);
    }

    return this;
  }

  async sync({
    collection,
    references,
    data,
    cursor,
    applyUpdate = true,
    runBulkOperation = true,
  }) {
    if (cursor) {
      while (await cursor.hasNext()) {
        const next = await cursor.next();
        await this.sync({ collection, references, data: next });
      }

      return Promise.resolve();
    }

    const docs = Array.isArray(data) ? data : [data];
    const refsConfig = this.references.get(collection);
    const refPropsList = Array.isArray(references) ? references : Object.keys(refsConfig);
    const refsIds = RefStore.getRefsIds({ docs, refPropsList });
    const refsList = await RefStore.getRefsList({ db: this.db, refsIds, refsConfig });
    const refsMap = refPropsList.reduce((acc, refPropr, index) => ({
      ...acc,
      [refPropr]: refsList[index],
    }), {});

    const bulk = this.db.collection(collection).initializeUnorderedBulkOp();

    let hasOperations = false;
    docs.forEach((doc) => {
      const updatePayload = RefStore.getDocUpdatePayload({ doc, refsConfig, refsMap });
      const hasUpdates = Object.keys(updatePayload).length;
      if (hasUpdates) {
        hasOperations = true;

        if (runBulkOperation) {
          bulk.find({ _id: doc._id }).updateOne({
            $set: updatePayload,
          });
        }

        if (applyUpdate) {
          Object.keys(updatePayload).forEach((key) => {
            set(doc, key, updatePayload[key]);
          });
        }
      }
    });

    if (hasOperations && runBulkOperation) {
      return bulk.execute();
    }

    return Promise.resolve();
  }

  async syncAll({ collections } = {}) {
    if (!collections) {
      collections = Array.from(this.references.getCollections()); // eslint-disable-line
    }

    return Promise.all(
      collections.map(
        (collection) => this.sync({ collection, cursor: this.db.collection(collection).find() })
      )
    );
  }

  onUpdate({ source, destination, refProperty, extractor, type, ns }) {
    this.events.on(operations.update, destination, async ({ query }) => {
      const refDocs = await this.db.collection(destination).find(query).toArray();
      const bulk = this.db.collection(source).initializeUnorderedBulkOp();

      refDocs.forEach((doc) => {
        const updateNS = type === types.one ? ns : `${ns}.${doc._id}`;

        bulk.find({
          [refProperty]: doc._id,
        }).update({
          $set: {
            [updateNS]: extractor(doc),
          },
        }, {
          multi: true,
        });
      });

      return bulk.execute();
    });
  }

  onRemove({ source, destination, refProperty, type, ns }) {
    this.events.on(operations.remove, destination, async ({ query }) => {
      const refDocs = await this.db.collection(destination).find(query, { _id: 1 }).toArray();
      const refDocsIds = refDocs.map(({ _id }) => _id);
      const updatePayload = {
        $unset: {},
      };

      if (type === types.one) {
        updatePayload.$unset[ns] = '';
        updatePayload.$unset[refProperty] = '';
      } else {
        refDocsIds.forEach((id) => {
          updatePayload.$unset[`${ns}.${id}`] = '';
          updatePayload.$pull = {
            [refProperty]: id,
          };
        });
      }

      return this.db.collection(source).update({
        [refProperty]: {
          $in: refDocsIds,
        },
      }, updatePayload, {
        multi: true,
      });
    });
  }

  notifyUpdate(collection, query) {
    return this.events.notify({
      operation: operations.update,
      collection,
      query,
    });
  }

  notifyRemove(collection, query) {
    return this.events.notify({
      operation: operations.remove,
      collection,
      query,
    });
  }
}
