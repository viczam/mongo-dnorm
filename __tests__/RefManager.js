import RefManager from '../src/RefManager';
import { MongoClient } from 'mongodb';

describe('Octobus', () => {
  let rm;
  let db;
  let products;
  let categories;

  beforeAll(() => (
    MongoClient.connect('mongodb://localhost:27017/mdnorm').then((_db) => {
      db = _db;
    })
  ));

  afterAll(() => {
    db.close();
  });

  beforeEach(() => (
    db.dropDatabase().then(async () => {
      rm = new RefManager(db);

      rm
        .add({
          source: 'Product',
          destination: 'Category',
          extractor: ({ name }) => ({ name }),
        })
        .add({
          source: 'Category',
          destination: 'Product',
          type: 'many',
          extractor: ({ name }) => ({ name }),
        });

      categories = (await db.collection('Category').insert([
        { _id: 1, name: 'category1' },
        { _id: 2, name: 'category2' },
        { _id: 3, name: 'category3' },
        { _id: 4, name: 'category4' },
      ])).ops;

      products = (await db.collection('Product').insert([
        { _id: 1, name: 'product1', categoryId: categories[0]._id },
        { _id: 2, name: 'product2', categoryId: categories[1]._id },
        { _id: 3, name: 'product3', categoryId: categories[3]._id },
        { _id: 4, name: 'product4', categoryId: categories[3]._id },
      ])).ops;

      await Promise.all([
        await db.collection('Category').update({
          _id: categories[0]._id,
        }, {
          $set: {
            productIds: [products[0]._id],
          },
        }),
        await db.collection('Category').update({
          _id: categories[1]._id,
        }, {
          $set: {
            productIds: [products[1]._id],
          },
        }),
        await db.collection('Category').update({
          _id: categories[3]._id,
        }, {
          $set: {
            productIds: [products[2]._id, products[3]._id],
          },
        }),
      ]);

      categories = await db.collection('Category').find().toArray();
    })
  ));

  it('stores the references configuration', () => {
    expect(rm.references).toMatchSnapshot();
  });

  describe('it should populate the cache after insert', () => {
    it('for type=one reference', async () => {
      await rm.notifyInsert('Product', products);
      const productsWithReferences = await db.collection('Product').find().toArray();
      expect(productsWithReferences).toMatchSnapshot();
    });

    it('for type=many reference', async () => {
      await rm.notifyInsert('Category', categories);
      const categoriesWithReferences = await db.collection('Category').find().toArray();
      expect(categoriesWithReferences).toMatchSnapshot();
    });
  });
});
