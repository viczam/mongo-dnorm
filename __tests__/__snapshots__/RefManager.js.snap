exports[`Octobus it should populate the cache after insert for type=many reference 1`] = `
Array [
  Object {
    "_id": 3,
    "name": "category3"
  },
  Object {
    "_id": 1,
    "_products": Object {
      "1": Object {
        "name": "product1"
      }
    },
    "name": "category1",
    "productIds": Array [
      1
    ]
  },
  Object {
    "_id": 2,
    "_products": Object {
      "2": Object {
        "name": "product2"
      }
    },
    "name": "category2",
    "productIds": Array [
      2
    ]
  },
  Object {
    "_id": 4,
    "_products": Object {
      "3": Object {
        "name": "product3"
      },
      "4": Object {
        "name": "product4"
      }
    },
    "name": "category4",
    "productIds": Array [
      3,
      4
    ]
  }
]
`;

exports[`Octobus it should populate the cache after insert for type=one reference 1`] = `
Array [
  Object {
    "_category": Object {
      "name": "category1"
    },
    "_id": 1,
    "categoryId": 1,
    "name": "product1"
  },
  Object {
    "_category": Object {
      "name": "category2"
    },
    "_id": 2,
    "categoryId": 2,
    "name": "product2"
  },
  Object {
    "_category": Object {
      "name": "category4"
    },
    "_id": 3,
    "categoryId": 4,
    "name": "product3"
  },
  Object {
    "_category": Object {
      "name": "category4"
    },
    "_id": 4,
    "categoryId": 4,
    "name": "product4"
  }
]
`;

exports[`Octobus stores the references configuration 1`] = `
Object {
  "Category": Object {
    "productIds": Object {
      "collection": "Product",
      "extractor": [Function extractor],
      "ns": "_products",
      "type": "many"
    }
  },
  "Product": Object {
    "categoryId": Object {
      "collection": "Category",
      "extractor": [Function extractor],
      "ns": "_category",
      "type": "one"
    }
  }
}
`;
